allow tests to run back to back

KauzClay opened the pull request:
Fixes #371 

I would like to be able to run the e2e tests multiple times without cleaning everything up. Particularly, I don't want to have to recreate the vcsim. 

From what I have seen, vcsim will remember the state of things. So if in the source test, we power off two vms, when we run again and try to power off those same vms, nothing will happen. No events will be sent, and the test will fail.

Same thing happens for the TestBindingGOVC. The tag and category will already exist, so those commands won't prompt any new events.

## Proposed Changes
- :broom: Update or clean up current behavior
- added helper functions to create/cleanup vcsim and secret
- use those helper functions in the tests
- added a checkpoint configmap to the CreateSource function. This will make sure the source goes back in time for events in case the test events happen before the source is ready.
- improved cleanup functions to also clean up completed job pods associated with listener and govc jobs

### Pre-review Checklist

<!-- If these boxes are not checked, you will be asked to complete these requirements or explain why they do not apply to your PR. -->

- [ ] **At least 80% unit test coverage** (I believe this does not apply, since I am modifying a test)
- [X] **E2E tests** for any new behavior
- [ ] **Docs** for any user-facing impact (I don't believe there is any user-facing impact)

**Release Note**

<!--
:page_facing_up: If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release note is needed.
-->

```release-note
N/A no user-facing impact
```

embano1 commented:
QQ: shouldn't we actually do cleanup after test, eg delete deployments, instead of manually resetting state?

KauzClay commented:
hi @embano1 , are you suggesting that the `vcsim` deployment be applied/deleted as part of the test too? 

As far as I can tell, the vcsim deployment isn't applied as part of the test. You have to deploy it manually before running the e2e, and therefore, delete it manually too if you want to rerun

embano1 commented:
> hi @embano1 , are you suggesting that the `vcsim` deployment be applied/deleted as part of the test too?

Well, IMHO that would fix your problem and avoid sharing state between runs.

> As far as I can tell, the vcsim deployment isn't applied as part of the test. You have to deploy it manually before running the e2e, and therefore, delete it manually too if you want to rerun

True, this repo still sets it up outside of the Go test framework. Two solutions:

- add a step in the workflow to delete any state, e.g. `vcsim`
- update the E2E tests to also manage the lifecycle of `vcsim` (which I typically do in other projetcts)

KauzClay commented:
okay great, that sounds good then. 

>update the E2E tests to also manage the lifecycle of vcsim (which I typically do in other projetcts)

do you have an example of where you do this elsewhere?

KauzClay committed:
add helper for creating vcsim

KauzClay committed:
create vcsim in source test

KauzClay committed:
manually create checkpoint

* it seems like when running the test, the events are being sent to
  vcsim before the vsphere source is ready. This means the source misses
  the events, and the test fails
* apparently, there is a default checkpoint window of 5 mins. However,
  that doesn't seem to be applying. This may need separate investigation
* creating a checkpoint configmap ourselves allows us to guarantee the
  source will ask vcsim for events from the past N minutes

KauzClay committed:
remove dead code

KauzClay committed:
create vcsim in all tests

KauzClay committed:
clean up job pods

* when the tests run, there are two lefover pods for the listener job and the govc job.
* this commit cleans those up, so upon finishing the test, all resources are cleaned up from the cluster

KauzClay committed:
run update-codegen.sh

KauzClay committed:
update git workflow

* now that e2e tests create and manage vcsim and secret, setup doesn't need to do it

embano1 commented on test/e2e/util.go:
```diff
@@ -201,6 +218,10 @@ func RunJobListener(t *testing.T, clients *test.Clients) (string, context.Cancel
 		}
+		err = clients.KubeClient.CoreV1().Pods(job.Namespace).DeleteCollection(context.Background(), metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", jobNameKey, name)})
+		if err != nil {
+			t.Errorf("Error cleaning up pod for Job %s", job.Name)
```
nit: since it's collection (same for other collection)

```suggestion
			t.Errorf("Error cleaning up pods for Job %s", job.Name)
```

embano1 commented on test/e2e/util.go:
```diff
@@ -7,24 +7,37 @@ package e2e
+const (
+	vcsim        = "vcsim"
+	ns           = "default"
+	vsphereCreds = "vsphere-credentials"
```
can you please use this const in `"--secret-ref", "vsphere-credentials",` too?

embano1 commented on test/e2e/util.go:
```diff
@@ -123,6 +136,10 @@ func RunJobScript(t *testing.T, clients *test.Clients, image string, command []s
 		if err != nil {
 			t.Errorf("Error cleaning up Job %s", job.Name)
 		}
+		err = clients.KubeClient.CoreV1().Pods(job.Namespace).DeleteCollection(context.Background(), metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", jobNameKey, job.Name)})
```
nit: can you please make `context.Background()` in all functions a variable?

embano1 commented on test/e2e/util.go:
```diff
@@ -319,5 +360,150 @@ func CreateSource(t *testing.T, clients *test.Clients, name string) context.Canc
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name:  vcsim,
+						Image: "vmware/vcsim:latest",
```
non-blocking: if we see `docker.io` `429`s we can always fall back to using `ko:` to build `vcsim` since it's already a vendored dependency.

embano1 reviewed (commented):
While reviewing this PR I saw that `RunJobListener()` calls `pkgtest.CleanupOnInterrupt` twice in the function. IMHO this is a bug if you look at how `pkgtest.CleanupOnInterrupt` works internally. Can you PTAL?

And in general I think the way we use `pkgtest.CleanupOnInterrupt` across the tests might not achieve the goal since multiple goroutines are registered, blocking for a `signal` and then racing to call `os.Exit(1)`

KauzClay committed:
refactor: address nits

KauzClay committed:
refactor: make context a variable

KauzClay commented:
@embano1 I addressed the nits, hopefully to what you were expecting :) 

Also, with regard to this comment...
> While reviewing this PR I saw that RunJobListener() calls pkgtest.CleanupOnInterrupt twice in the function. IMHO this is a bug if you look at how pkgtest.CleanupOnInterrupt works internally. Can you PTAL?
>
> And in general I think the way we use pkgtest.CleanupOnInterrupt across the tests might not achieve the goal since multiple goroutines are registered, blocking for a signal and then racing to call os.Exit(1)

The way I'm reading it, it seems like calling it multiple times is okay. It looks like `cleanup.funcs = append(cleanup.funcs, cleanupFunc)` is just adding them to the slice of functions.

Initially I got hung up on the `cleanup.once.Do(waitForInterrupt)` line, but I believe the `once.Do()` means that every call after that will skip over that line, so I think there is only 1 goroutine actually blocking on a call to `os.Exit(1)`.

I tested it out here: https://gist.github.com/KauzClay/abf3fe664e6867bca90dc16c8a11f594
And this is what I saw:
```
❯ go run main.go      
doing stuff
sleeping, cancel me
^CI am cleanup function 3
I am cleanup function 2
I am cleanup function 1
exit status 
```

embano1 reviewed (approved):
LGTM, thx mate!

And regarding CleanupOnInterrupt: I don't know which code I looked at when I wrote this. Just checked src again and it definitely does not call exit more than once. All good 👍

gabo1208 commented:
Let me know if this is ready to review al lgtm :)
